fix: properly request docker images from docker API

lengau opened the pull request:
Fixes #1760, reverts #1761

CRAFT-3148

lucabello commented:
`upload-resource` fails with `DockerException()` on 3.2 (but works on 2.7) for me

Command:
```bash
charmcraft upload-resource blackbox-exporter-k8s blackbox-exporter-image --image="docker://quay.io/prometheus/blackbox-exporter:v0.24.0" --format=json
```

Logs: https://pastebin.canonical.com/p/GmrtZCvjRV/

lengau commented:
Thanks @lucabello! I tested it with this case and it failed, but I've modified the PR and added a fix that worked for me locally. Could you test it from the `edge/pr-1854` channel to ensure it wasn't a fluke?

lucabello commented:
Confirmed it works from your branch (`rev 3.2.0.post9+g2460b772`). Thanks!

lengau commented on charmcraft/services/image.py:
```diff
@@ -54,14 +55,17 @@ class ImageService(craft_application.AppService):
+        try:
+            self._docker = docker.from_env()
+        except docker.errors.DockerException:
+            self._docker = None
```
Yeah that's a good point, thanks. I'll just log it in debug level since the most common use case is that docker isn't installed.

lengau commented on snap/snapcraft.yaml:
```diff
@@ -187,7 +187,7 @@ parts:
     build-packages:
       - golang-go
     build-environment:
-      - CGO_ENABLED: 0
```
https://github.com/canonical/snapcraft/issues/4978

tigarmo commented on charmcraft/const.py:
```diff
@@ -98,6 +98,20 @@ def __str__(self) -> str:
         return str(self.value)
 
 
+GO_ARCH_TO_CHARM_ARCH = {
```
what about the "variant"? https://github.com/canonical/rockcraft/blob/main/rockcraft/oci.py#L102

lengau replied:
We don't worry about variants, as charmhub doesn't. If the OCI contains any image that's compatible with the arch, we tell charmhub that we support that arch.
